### PR TITLE
CUDA: Fix linking with PTX when compiling lazily

### LINKED ISSUE #7607: CUDA: Linking PTX only works for eager compilation
Making the following change to the PTX linking test:

```diff
diff --git a/numba/cuda/tests/cudadrv/test_linker.py b/numba/cuda/tests/cudadrv/test_linker.py
index 0a4dc1f8a..87833906c 100644
--- a/numba/cuda/tests/cudadrv/test_linker.py
+++ b/numba/cuda/tests/cudadrv/test_linker.py
@@ -73,7 +73,7 @@ class TestLinker(CUDATestCase):
 
         link = os.path.join(os.path.dirname(__file__), 'data', 'jitlink.ptx')
 
-        @cuda.jit('void(int32[:], int32[:])', link=[link])
+        @cuda.jit(link=[link])
         def foo(x, y):
             i = cuda.grid(1)
             x[i] += bar(y[i])
```

results in the following failure:

```
======================================================================
ERROR: test_linking (numba.cuda.tests.cudadrv.test_linker.TestLinker)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 2707, in complete
    driver.cuLinkComplete(self.handle, byref(cubin), byref(size))
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 319, in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 384, in _check_ctypes_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [300] Call to cuLinkComplete results in CUDA_ERROR_INVALID_SOURCE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/devices.py", line 232, in _require_cuda_context
    return fn(*args, **kws)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudadrv/test_linker.py", line 84, in test_linking
    foo[1, 1](A, B)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 723, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 911, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 919, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 1089, in compile
    kernel.bind()
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 466, in bind
    self._codelibrary.get_cufunc()
  File "/home/gmarkall/numbadev/numba/numba/cuda/codegen.py", line 201, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "/home/gmarkall/numbadev/numba/numba/cuda/codegen.py", line 178, in get_cubin
    cubin_buf, size = linker.complete()
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 2709, in complete
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [300] Call to cuLinkComplete results in CUDA_ERROR_INVALID_SOURCE
error   : Undefined reference to 'bar' in '<cudapy-ptx>'
```

I note that the PTX is never linked (no call to `Linker.add_file()` is reached, but haven't investigated further.

### PR DESCRIPTION
**Pull request by gmarkall**
When no signature was provided, the `link` kwarg was accidentally dropped in the wrapper function calling `jit` inside the decorator.

Also includes the removal of some un-needed decorators in the linker test.

Fixes #7607.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/tests/cudadrv/test_linker.py [hunk 1]
```diff
@@ -66,14 +66,18 @@ def test_linker_basic(self):
         linker = Linker.new()
         del linker
 
-    @require_context
-    def test_linking(self):
+    def _test_linking(self, eager=False):
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudadrv/test_linker.py:69]:
```suggestion
    def _test_linking(self, eager):
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. One minor suggestion else looks good.

**Comment by gmarkall** [comment]
@stuartarchibald Thanks for the review - change applied.

gpuci run tests

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_111`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_111`.

Passed.

**Pull request merged**